docs/adr/2023-12-06_029-use-cbor-for-tx.md

```diff
@@ -0,0 +1,49 @@
+  * We have specified this format as part of [Hydra API](https://github.com/input-output-hk/hydra/blob/b2dc5a0da4988631bd2c1e94b66ba6217d5db595/hydra-node/json-schemas/api.yaml#L1473)
+* These instances appear in a few places as part of Hydra API:
+  * In the [ServerOutput](https://github.com/input-output-hk/hydra/blob/b2dc5a0da4988631bd2c1e94b66ba6217d5db595/hydra-node/src/Hydra/API/ServerOutput.hs#L51) sent by the node to clients
+  * In the [HydraNodeLog](https://github.com/input-output-hk/hydra/blob/b2dc5a0da4988631bd2c1e94b66ba6217d5db595/hydra-node/src/Hydra/Node.hs#L122) as part of Hydra's logging output
+  * In the [StateChanged](https://github.com/input-output-hk/hydra/blob/b2dc5a0da4988631bd2c1e94b66ba6217d5db595/hydra-node/src/Hydra/HeadLogic/Outcome.hs#L46) events which are persisted and allow hydra-node to restart gracefully after stopping
+* In other places the hydra-node produces, expects, or accepts a CBOR-encoded transaction:
+  * In the [Network.Message](https://github.com/input-output-hk/hydra/blob/b2dc5a0da4988631bd2c1e94b66ba6217d5db595/hydra-node/src/Hydra/Network/Message.hs#L20) exchanged between the nodes
+  * In the [ClientInput](https://github.com/input-output-hk/hydra/blob/b2dc5a0da4988631bd2c1e94b66ba6217d5db595/hydra-node/src/Hydra/API/ClientInput.hs#L9) from clients submitting `NewTx` commands
+  * In the [HTTPServer](https://github.com/input-output-hk/hydra/blob/b2dc5a0da4988631bd2c1e94b66ba6217d5db595/hydra-node/src/Hydra/API/HTTPServer.hs#L297) API
+* Note that in the latter 2 cases, the hydra-node _accepts_ a hex-CBOR-encoded _JSON string_ to represent a transaction and this particular case is handled directly in the [FromJSON](https://github.com/input-output-hk/hydra/blob/b2dc5a0da4988631bd2c1e94b66ba6217d5db595/hydra-node/src/Hydra/Ledger/Cardano/Json.hs#L388) instance for transactions where 3 different representations are even accepted:
+  * JSON object detailing the transaction
+  * A JSON string representing CBOR-encoding of a transaction
+  * Or a `TextEnvelope` which wraps the CBOR transaction in a simple JSON object
+* Using JSON-based representation of Cardano transactions is problematic because:
+  * The representation we are providing is not _canonical_ nor widely used, and therefore require maintenance when the underlying cardano-ledger API changes
+  * **More importantly** the JSON representation contains a `txId` field which is computed from the CBOR encoding of the transaction. When this encoding changes, the transaction id changes even though no other part of the transaction has changed. This implies that we could send and receive transactions with incorrect or inconsistent identifiers.
```

Comment:
I think the problem arose when our decoder was producing a "valid, but not identical" value for the Tx. Which can happen when we construct a ledger transaction without using their memoizing CBOR decoders, but the normal data constructors.

Not sure if this is the same thing as you mention ("encoding changes").

Reply:
Do you mean `TxId` or `Tx`? I understood the problem to be that
1. we decode a JSON representation of a transaction, giving us an in-memory data structure
2. to compute the `txId` for this transaction, we use whatever function the ledger provides which computes it from the CBOR encoding of the transaction
3. given this encoder has changed, the transaction id is not the same as the one we initially computed and stored in the JSON representation. This would of course not have been the case had we kept the CBOR encoding of the transaction and not its JSON encoding, because the decoder must be backward compatible

Did I miss something?

Reply:
I mean `Tx`. It think you are right in the cause of effects. But I think it would have worked with CBOR because the CBOR encoder of a previously decoded `Tx` is to just output the original memoized bytes. That allows the ledger to produce byte-for-byte the same CBOR again if there are multiple valid ways to encode some Haskell value.

Reply:
> the CBOR encoder of a previously decoded Tx is to just output the original memoized bytes. That allows the ledger to produce byte-for-byte the same CBOR again even if there are multiple valid ways to encode some Haskell value.

This is "their" way of handling backward-compatibility in the face of lack of versioning, which they have now added to their CBOR content AFAIU. The usual way is to version the encoded data and maintain compatibility by keeping track of how to convert between versions. 

> some version bump changed the "canonical" encoding for something in the transaction

That's my understanding too.

Anyway, I think we agree, do you want to change how it's formulated in the ADR?

Reply:
It's fine how it is phrased (it's vague enough) and the details do not really help the ADR. I did resolve this conversation already, so fine by me :)